DRAFT - mdx for Synthetic Data Creation lung images

kfahn22 opened the pull request:
This is a first draft for the section on generating lung images with DCGAN. Note that I am not thrilled with the quality of the lung images. I have started looking into  Medfusion (which uses to diffusors) to see if the images can be improved, although not sure about the time-frame to implement.

![lung_images](https://github.com/johko/computer-vision-course/assets/65121394/0bc09dde-fa0d-4198-ba95-2ce9703a9585)

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+
+## DCGAN (Deep Convolutional Generative Adversarial Network):
+
+DCGAN was proposed in [Unsupervised Representation Learning With Deep Convolutional Generative Adversarial Networks](https://arxiv.org/abs/1511.06434) by Radford, et al. and is the model
```
```suggestion
DCGAN was proposed in [Unsupervised Representation Learning With Deep Convolutional Generative Adversarial Networks](https://arxiv.org/abs/1511.06434) by Radford, et al. and is the
```

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+## DCGAN (Deep Convolutional Generative Adversarial Network):
+
+DCGAN was proposed in [Unsupervised Representation Learning With Deep Convolutional Generative Adversarial Networks](https://arxiv.org/abs/1511.06434) by Radford, et al. and is the model
+most often used for medical imaging. We are going to use it to generate synthetic lung
```
```suggestion
most often used model for medical imaging. We are going to use it to generate synthetic lung
```
Can you remove the newlines?

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+
+Here are the basic steps:
+
+_Data Collection:_
```
```suggestion
**Data Collection**
```
We can make smallest headers with bold

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+Here are the basic steps:
+
+_Data Collection:_
+First, We need to obtain a dataset of real lung images.
```
```suggestion
First, we need to obtain a dataset of real lung images.
```

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+
+_Data Collection:_
+First, We need to obtain a dataset of real lung images.
+We will be downloading the [Chest X-Ray Images (Pneumonia)](https://www.kaggle.com/datasets/paultimothymooney/chest-xray-pneumonia) dataset from [Kaggle](https://www.kaggle.com).
```
If you feel like taking the extra step, you could put this in hf.co/hf-vision hugging face organization as a dataset repository and load from there 😊

kfahn22 replied:
Done

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+We will be downloading the [Chest X-Ray Images (Pneumonia)](https://www.kaggle.com/datasets/paultimothymooney/chest-xray-pneumonia) dataset from [Kaggle](https://www.kaggle.com).
+You can learn more about the dataset by looking at the Datacard on Kaggle.
+
+_Data Preprocessing:_
```
```suggestion
**Data Preprocessing**
```

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+_Data Preprocessing:_
+We will preprocess the lung images by resizing and normalizing the pixel values.
+
+_DCGAN Architecture:_
```
```suggestion
**Create the DCGAN Architecture**
```

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+
+We will use the following model architecture:
+
+Generator(
```
```suggestion
```bash
Generator(
```
Maybe what's even better is to explain the architecture verbally like in [here](https://github.com/johko/computer-vision-course/pull/57/files#diff-e242ea8730e66737602f9e9bfb772b553da7c9c8846199d2c9552200a3340222R13)

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+)
+)
+
+## Training:
```
```suggestion
## Training
```

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+
+## Resources and Further Reading
+
+[A Novel COVID-19 Detection Model Based on DCGAN and Deep Transfer Learning](https://www.sciencedirect.com/science/article/pii/S1877050922007463)
```
Maybe you could put it in list using - for better readability

merveenoyan reviewed (commented):
Reviewed the markdown, now off to the notebook 😊

merveenoyan commented:
I think using kaggle adds quite a lot of complexity. Overall the notebook is pretty informative. I'd like to have @MKhalusova add her views too!

MKhalusova commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+## How did we do?
+
+We can see that some of the synthetic lung images look good, but some are hazy. Researchers generating synthetic medical images typically
+employee a "human in the middle"--in this case, a expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with
```
```suggestion
employ a "human in the middle"--in this case, an expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with
```

MKhalusova commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+
+We can see that some of the synthetic lung images look good, but some are hazy. Researchers generating synthetic medical images typically
+employee a "human in the middle"--in this case, a expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with
+with the real data to train the model.
```
```suggestion
the real data to train the model.
```

MKhalusova commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+employee a "human in the middle"--in this case, a expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with
+with the real data to train the model.
+
+![lung-images]()
+
+[[Link to Notebook]]
```
Links are missing.

MKhalusova commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+[Deepfake Image Generation for Improved Brain Tumor Segmentation](https://aps.arxiv.org/abs/2307.14273)
+[GANs for Medical Image Synthesis: An Empirical Study](https://arxiv.org/abs/2105.05318)
+[Medfusion Github repo](https://github.com/mueller-franzes/medfusion)
+[Medical image editing in the latent space of Generative Adversarial Networks](https://pdf.sciencedirectassets.com/777072/1-s2.0-S2666521221X00026/1-s2.0-S2666521221000168/main.pdf?X-Amz-Security-Token=IQoJb3JpZ2luX2VjEBUaCXVzLWVhc3QtMSJGMEQCIBa2IzG%2BGD7Hjv%2FP20jDKyseswGT785LUQTMmFQjT9ntAiAt%2ByfhQJqJAI2NCOKnVzH%2BxMlTWf2x2uIw2i2OLFOz8yqzBQh%2BEAUaDDA1OTAwMzU0Njg2NSIM9ECuDI1xe%2FlWShlYKpAFeTDTJ ...
```
It looks like the link can be much shorter: 
`https://www.sciencedirect.com/science/article/pii/S2666521221000168?ref=pdf_download&fr=RR-2&rr=833e48fa5e777142`

MKhalusova reviewed (commented):
Thank you for working on this! I have left some comments on what can be improved.

johko commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,115 @@
+## DCGAN (Deep Convolutional Generative Adversarial Network):
+
+DCGAN was proposed in [Unsupervised Representation Learning With Deep Convolutional Generative Adversarial Networks](https://arxiv.org/abs/1511.06434) by Radford, et al. and is the model
+most often used for medical imaging. We are going to use it to generate synthetic lung
```
Coming from an academic background: Do we have proof that it is the most widely used model for that or is that just an educated assumption?
If there is some source that states it, it would be nice to include it.

kfahn22 replied:
I appreciate the comment since I also have an academic background. The truth is I don't remember if I read that or that was just my impression after reading a bunch of studies, so I have edited the sentence.

johko reviewed (commented):
Nice writing again Kathy! 
I have not much to add to the review from Merve and Maria for the mdx, will have a look at the notebook now.

kfahn22 commented:
<p>done</p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1848233752'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>thanks fixed</p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1850171576'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>Have see the misspelling on preprocess.  Are there others?</p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1857390945'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>I have to admit I really struggled trying to show the images after they were downloaded from the Hub.  I am guessing there must be an easy way to do it, but I haven't figured it out.  I can filter them to a folder, but that cell was taking A REALLY LONG TIME and I decided it wasn't worth it.  I have settled on just showing one image.  If you can tell me how to edit I will update.</p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1852889747'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>I am sorry, I am not sure what this is referring to.  Please let me know if I still need to fix.</p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1857390989'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>Sorry, I am not exactly sure what you are saying.  This notebook ran in High-Ram (although it took several hours)- is that what you are referring to? </p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1852889714'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
I have substantially edit the notebook.   I tried to address the comments, but I might have missed some things.  Please let me know what I still  need to update.   Note that the results seem to be highly dependent on the seed value, which I am still finding hard to believe.  At any rate, I changed the seed value to 22 (as much as a Hitchhiker's nerd as I am, 42 led to a poor outcome).  Also, I am wondering if the model is suffering from mode collapse since at least to me the verified images look similar.  I edited the mdx file to discuss this.

kfahn22 commented:
I have also added links to the training GIF and image showing the verified synthetic Xray images.

MKhalusova commented:
<p>I agree with Johannes -  it would be useful to add a note specifying what setup a learner might need to run this notebook (e.g. Google Colab with High RAM A100). </p><p>This is especially important when the free version of Colab is not enough to run the notebook (I think it's only T4, and no high-RAM). </p><p>This is something that would be great to add to all the notebooks (but especially the ones that won't run on the free tier)</p><p> </p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1852889714'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

MKhalusova commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,111 @@
+
+![lung-images](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/dcgan-lung_images/verified.png)
+
+We can see that most of the synthetic lung images look good, but a couple are hazy. There are some important things to mention. First, researchers generating synthetic medical images typically employee a "human in the middle"--in this case, a expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with the real data to train the model. Second, ...
```
```suggestion
We can see that most of the synthetic lung images look good, but a couple are hazy. There are some important things to mention. First, researchers generating synthetic medical images typically employ a "human in the middle"--in this case, an expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with the real data to train the model. Second, even though the generated images appear to look OK (at least to an amateur), they do appear to be very similar. This is another known issue with GANs - they suffer from "mode collapse." Essentially, this happens when the generator starts producing the same type of output. Think of it as someone who has gotten a lot of praise for making chocolate chip cookies and, therefore, gets _really, really_ good at making those cookies but can't make any other type of cookies.
```

MKhalusova reviewed (approved):
Left a minor nit for the mdx file, the rest of it looks great! Will take a separate look at the notebook.

kfahn22 commented:
<p>I have changed the code to display 4 normal images and 4 pneumonia images, but they are all big.  </p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1860768910'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>Fixed</p><p> </p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1860768950'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>I thought about pushing the model to the Hub, but I wondered whether the model was good enough.  I am exploring MedFusion, which first trains a VAE  or VAEGAN and then  trains further with a Diffusor.  </p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1860769070'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1857390973'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1857390989'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kfahn22 commented:
<p>I have changed the naming and added an explanation.</p>

---

 <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/105/discussion/#comment-1860769139'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

merveenoyan commented on chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx:
```diff
@@ -0,0 +1,111 @@
+
+We can see that most of the synthetic lung images look good, but a couple are hazy. There are some important things to mention. First, researchers generating synthetic medical images typically employ a "human in the middle"--in this case, a expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with the real data to train the model. Second, e ...
+
+[[Link to Notebook]]
```
You can put the link here (internally, e.g. (./name-of-notebook.ipynb)

kfahn22 replied:
Is it just the name and not the relative path? (That is what I have added.)

merveenoyan replied:
relative path itself yep

merveenoyan reviewed (approved):
We can get this merged!

kfahn22 commented:
Sorry this took me a while.  I  realized that in changing the code to download the dataset from the hub, I introduced a bug.  

I think I have addressed all of the comments on the notebook.  I have added a discussion on preprocessing, training, the plot, and "good" images.  I also added a link to the notebook. 

Please let me know if there is anything else--thank you!
